### ISSUE TITLE
Automated regression tests

### ISSUE THREAD

**Issue by DirtyHairy** [stella-emu/stella#90]
I have been thinking about how to automate regression testing --- that'd give me a lot more confidence when fiddling with timing details inside the TIA :smile: My preliminary plan is:

* Extend Stella with an option to fix the random seed and automatically do a screenshot after N frames, then quit.
* Write a test harness that automatically iterates over a set of know goods and compares the screenshots. The ROMs could be referenced by SHA1 and taken from a downloaded copy Romhunter's collection (and possibly some testcases in the repo).

I'd propably do the harness in Golang --- I'd be working much faster there than in C++, and paralellizing tests would basically come for free. @sa666666 , @thrust26 , what do you think?

**Comment by sa666666** [comment]
There is already an (undocumented) commandline argument to start Stella, take a snapshot and then exit: 'takesnapshot'.  It basically runs the emulation for 30 frames, then takes a snapshot and exits.  The code is at line 130 of src/common/main.cxx.  Perhaps this can be extended.

As for your other suggestions, I see no problems with doing them.

**Comment by DirtyHairy** [comment]
Thanks for the pointer, that's a good point to start. I'll try and get something rigged up then.

**Comment by thrust26** [comment]
I am all for automatized tests. Would you check in the code into GitHub too?

For the tests, maybe comparing hashcodes of screenshot would do too. That would save some time. 

I would suggest, that you take multiple screenshots and press some keys in between like SELECT, RESET and FIRE. And maybe some random joystick movements too. Or do you want to define this for each ROM individually?

**Comment by DirtyHairy** [comment]
Definitely. I guess I'll create another repo under the stella organization for the harness and associated definitions. I have also been considering using a hash for comparision, but I think I'll start of with the full screenshot first --- I think that the slowest part is generating the shot, and using the full screenshot would also allow to generate an image with only the differences. If performance turns out to be an obstacle, I can always move to hashes.

In order to keep things simple, I want to start out by taking a single shot after a bunch of frames. That's what I'd like to get running before the next prerelease, but I have also thought about scripting the tests and injecting events after a predefined frame count as a future improvement. Random input is a nice idea, too :smirk:

**Comment by sa666666** [comment]
@DirtyHairy and @thrust26, I suppose since we haven't added automated testing at this point, we are never going to.  So maybe we can close this one?

**Comment by DirtyHairy** [comment]
Hm, I still think this would be really useful, and I still am willing to work on this, although I understand I have a backlog of stuff I want to work on and never did so far 😏 I would like to keep it open.
